File: 2024/14_toiletry_and_robots.py
import numpy as np
from collections import defaultdict
import matplotlib.pyplot as plt

example = """p=0,4 v=3,-3
p=6,3 v=-1,-3
p=10,3 v=-1,2
p=2,0 v=2,-1
p=0,0 v=1,3
p=3,0 v=-2,-2
p=7,6 v=-1,-3
p=3,0 v=-1,-2
p=9,3 v=2,3
p=7,3 v=-1,2
p=2,4 v=2,-3
p=9,5 v=-3,-3"""

MAX_X = 11
MAX_Y = 7

dat = open('data/14.dat', 'rt').read()

def parse(inp):
    paths = []
    for line in inp.splitlines():
        p, v = [x.split('=')[1] for x in line.split()]
        p = tuple(int(x) for x in p.split(','))
        v = tuple(int(x) for x in v.split(','))
        paths.append((p, v))

    return paths

def convert_paths_to_numpy(paths):
    p0 = np.zeros((len(paths), 2), dtype=int)
    v = np.zeros((len(paths), 2), dtype=int)

    for i, (p, v_) in enumerate(paths):
        p0[i] = p
        v[i] = v_

    return p0, v

def positions_in_step(p0, v, step, max_x=MAX_X, max_y=MAX_Y):
    pos = p0 + step * v
    np.mod(pos, [max_x, max_y], out=pos)
    return pos

def count_quadrants(positions, max_x=MAX_X, max_y=MAX_Y):
    ctr = defaultdict(int)
    for pos in positions:
        rel_pos = pos - [max_x // 2, max_y // 2]
        if 0 not in rel_pos:
            quadrant = np.sign(rel_pos)
            ctr[tuple(quadrant)] += 1

    return np.asarray(list(ctr.values()))

def solve1(inp, max_x=MAX_X, max_y=MAX_Y):
    p0, v = convert_paths_to_numpy(parse(inp))
    pos = positions_in_step(p0, v, step=100, max_x=max_x, max_y=max_y)
    quadrants = count_quadrants(pos, max_x, max_y)
    print(np.prod(quadrants))

def convert_to_matrix(positions, max_x=MAX_X, max_y=MAX_Y):
    matrix = np.zeros((max_y, max_x), dtype=int)
    for pos in positions:
        matrix[pos[1], pos[0]] += 1

    return matrix

def solve2(inp, max_x=MAX_X, max_y=MAX_Y, num_steps=100):
    # Candidate with patterns are 12 and 35:
    p0, v = convert_paths_to_numpy(parse(inp))

    for k in range(num_steps):
        p = positions_in_step(p0, v, step=k, max_x=max_x, max_y=max_y)

        # This here is one condition, which I don't know why it works.
        # Why is the assumption that no robots are stacked true?
        # I guess that is just a lucky break as that is how the input is created.
        # In order to run the following condition, you have to add M = convert_to_matrix(p, max_x, max_y)
        # to the code above.
        #if M.max() == 1:
        #    plt.matshow(M)
        #    plt.title(k)
        #    plt.show()

        # But another condition makes sense: Every horizontal pattern repeats with every max_y lines
        # and every vertical pattern repeats with every max_x columns. So if at any given step k,
        # k is divisible by the respective frequency, then we have a candidate.
        # Before checking for divisibility, however, we have to subtract the respective
        # offset. In this case 12 for the horizontal pattern and 35 for the vertical pattern.
        if (k - 12) % max_y == 0 and (k - 35) % max_x == 0:
            M = convert_to_matrix(p, max_x, max_y)
            print(f'Solution for part 2: {k}')
            plt.matshow(M)
            plt.title(k)
            plt.show()


solve1(example)
solve1(dat, max_x=101, max_y=103)

solve2(dat, max_x=101, max_y=103, num_steps=10000)

File: 2024/14_toiletry_and_robots.jl
using DataStructures
using Base

example = split("p=0,4 v=3,-3
p=6,3 v=-1,-3
p=10,3 v=-1,2
p=2,0 v=2,-1
p=0,0 v=1,3
p=3,0 v=-2,-2
p=7,6 v=-1,-3
p=3,0 v=-1,-2
p=9,3 v=2,3
p=7,3 v=-1,2
p=2,4 v=2,-3
p=9,5 v=-3,-3", '\n')


mutable struct Robot
    x::Int
    y::Int
    vx::Int
    vy::Int
    xmax::Int
    ymax::Int
end
function move(r::Robot)::Robot
    r.x += r.vx
    r.y += r.vy

    if r.x < 1
        r.x += r.xmax
    elseif r.x > r.xmax
        r.x -= r.xmax
    end

    if r.y < 1
        r.y += r.ymax
    elseif r.y > r.ymax
        r.y -= r.ymax
    end
    return r
end
function get_quadrant(r::Robot)::Int
    #= 
    This returns a quadrant number of 1 to 4, starting with the top left and going clockwise.
    If an item is on the center cross, we return the value 0.
    =#
    if r.x == r.xmax ÷ 2 + 1 || r.y == r.ymax ÷ 2 + 1
        return 0
    end

    # Maybe a lookup would be more efficient and elegant and readable and better?
    #=
    if r.x ≤ r.xmax ÷ 2
        if r.y ≤ r.ymax ÷ 2
            return 1
        else
            return 2
        end
    else
        if r.y ≤ r.ymax ÷ 2
            return 3
        else
            return 4
        end
    end
    =#

    is_up = r.x ≤ r.xmax ÷ 2
    is_left = r.y ≤ r.ymax ÷ 2

    lookup = Dict(
        [((true, true), 1), ((true, false), 2), ((false, true), 3), ((false, false), 4)]
    )

    return lookup[(is_up, is_left)]
end


function parse_inp(inp, xmax=7, ymax=11)::Vector{Robot}
    #=
    Please note that we want to stick to the convention that x = axis 1 = going down,
    so we have to switch the values during parsing.
    Also, the values given are zero-based and Julia works one-based, so we have to add the offset.
    =#
    robots = Vector{Robot}()
    for line in inp
        y, x, vy, vx = [parse(Int, x.match) for x in eachmatch(r"-?\d+", line)]
        push!(robots, Robot(x + 1, y + 1, vx, vy, xmax, ymax))
    end
    return robots
end


function step(robots::Vector{Robot}, N::Int)::Vector{Robot}
    for k in range(1, N)
        robots = [move(r) for r in robots]
    end
    return robots
end


function score(robots::Vector{Robot})::Int
    quadrantcounts = DefaultOrderedDict{Int, Int}(0)

    for robot in robots
        quadrantcounts[get_quadrant(robot)] += 1
    end

    return reduce(*, [x for (k, x) in quadrantcounts if k > 0])
end


function print_robots(robots)
    # Use responsibly!
    xmax, ymax = robots[1].xmax, robots[1].ymax

    s = [["." for _ in range(1, ymax)] for _ in range(1, xmax)]

    positions = DefaultDict{Tuple{Int, Int}, Int}(0)
    for robot in robots
        positions[(robot.x, robot.y)] += 1
    end

    for (k, v) in positions
        s[k[1]][k[2]] = string(v)
    end

    println(join([join(line) for line in s], "\n"))
end


@time begin
    # Part 1:
    robots = parse_inp(readlines("data/14.dat"), 103, 101)
    robots = step(robots, 100)
    println("Part 1: ", score(robots))

    # Part 2:
    #=
        This part relies on manual analysis.

        You first have to look through e.g. the first 100 robot moves to realise that there are vertical and horizontal patterns showing up.
        I found these for my data at step 12 for the horizontal pattern and index 35 for the vertical pattern.
        Consider e.g. the horizontal pattern, which means that every item is in a sorted position horizontally, but not vertically.
        That means whenever the robots did a full horizontal cycle, we would have another chance at matching the horizontal and the vertical pattern.
        A cycle for the horizontal pattern takes xmax steps.
    
        This means we are looking for an index where both cycles fall together.
    =#
    local robots = parse_inp(readlines("data/14.dat"), 103, 101)
    xmax, ymax = robots[1].xmax, robots[2].ymax
    for k in range(1, 100000)
        robots = step(robots, 1)
        if (k - 12) % xmax == 0 && (k - 35) % ymax == 0
            println("Part 2: ", k)
            print_robots(robots)
            break
        end
    end

end